panoramapublic/src/org/labkey/panoramapublic/chromlib/ChromLibStateManager.java

```diff
@@ -0,0 +1,247 @@
+    static final DoubleHeader H_PREC_MZ = new DoubleHeader(PREC_MZ);
+    static final IntegerHeader H_PREC_CHARGE = new IntegerHeader(PREC_CHARGE);
+    static final StringHeader H_PREC_STATE = new StringHeader(PREC_STATE);
+    static final StringHeader H_PREC_MOD_SEQ = new StringHeader(PREC_MOD_SEQ);
+    static final StringHeader H_PREC_MOL_CUSTOM_ION_NAME = new StringHeader(PREC_MOL_CUSTOM_ION_NAME);
+    static final StringHeader H_PREC_MOL_ION_FORMULA = new StringHeader(PREC_MOL_ION_FORMULA);
+    static final DoubleHeader H_PREC_MOL_MASS_MONOISOTOPIC = new DoubleHeader(PREC_MOL_MASS_MONOISOTOPIC);
+    static final DoubleHeader H_PREC_MOL_MASS_AVERAGE = new DoubleHeader(PREC_MOL_MASS_AVERAGE);
+
+    static final List<Header<?>> PROTEIN_LIB_HEADERS = List.of(H_RUN, H_RUN_STATE, H_PEP_GRP, H_PEP_GRP_SEQ_ID, H_PEP_GRP_STATE);
+    static final List<String> PROTEIN_LIB_COLS = PROTEIN_LIB_HEADERS.stream().map(h -> h._name).collect(Collectors.toList());
+    static final List<Header<?>> PEP_LIB_HEADERS = List.of(H_RUN, H_RUN_STATE, H_PEP_GRP, H_PEP_GRP_SEQ_ID, H_PEP_GRP_STATE,
+            H_PREC_MZ, H_PREC_CHARGE, H_PREC_STATE,
+            H_PREC_MOD_SEQ, H_PREC_MOL_CUSTOM_ION_NAME, H_PREC_MOL_ION_FORMULA, H_PREC_MOL_MASS_MONOISOTOPIC, H_PREC_MOL_MASS_AVERAGE);
+    static final List<String> PEP_LIB_COLS = PEP_LIB_HEADERS.stream().map(h -> h._name).collect(Collectors.toList());
+
+    public static class Header<T>
+    {
+        final String _name;
+        final Class<T> _type;
+
+        private Header(String name, Class<T> type)
+        {
+            _name = name;
+            _type = type;
+        }
+    }
+
+    public final static class StringHeader extends Header<String>
+    {
+        private StringHeader(String name)
+        {
+            super(name, String.class);
+        }
+    }
+
+    public final static class IntegerHeader extends Header<Integer>
+    {
+        private IntegerHeader(String name)
+        {
+            super(name, Integer.class);
+        }
+    }
+
+    public final static class DoubleHeader extends Header<Double>
+    {
+        private DoubleHeader(String name)
+        {
+            super(name, Double.class);
+        }
+    }
+
+    public void copyLibraryState(Container sourceContainer, Container targetContainer, Logger log, User user) throws ChromLibStateException
+    {
+        Path chromLibDir = getChromLibDir(targetContainer);
+        Path chromLibExportFile = chromLibDir.resolve(FileUtil.makeFileNameWithTimestamp("chrom_lib_state_export_" + sourceContainer.getRowId(), "tsv"));
+        ChromLibStateExporter.exportLibState(chromLibExportFile.toFile(), sourceContainer, user, log);
+        ChromLibStateImporter.importLibState(chromLibExportFile.toFile(), targetContainer, user, log);
+    }
+
+    public static Path getChromLibDir(Container container)
+    {
+        PipeRoot root = PipelineService.get().getPipelineRootSetting(container);
+        assert root != null;
```

Comment:
Optional - I've never found asserts like this very helpful. On dev machines (where asserts are enabled) it would fail on line 115 with no message. On prod servers it would fail with an NPE on the next line. There's no real difference, especially now that Java NPEs tell you what variable was null